### PR TITLE
Sync and async modes

### LINKED ISSUE #52: Add sync method for result compile
### Async
```js
posthtml()
    .use(plugin)
    .process(html/*, options */)
    .then(function(result) {
        // result.html
        // result.tree
    });
```

### Sync
```js
var html = posthtml()
    .use(plugin)
    .process(html/*, options */)
    .html; // or .tree
```

### PR DESCRIPTION
**Pull request by zxqfox**
- implement truly sync and async modes
- tests for them
- rework tests for Plugins

/cc @voischev @awinogradov 

Fixes #52

### CODE REVIEW

File: lib/posthtml.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import { toTree, toHtml } from './parser.js';
 import api from './api.js';
 
+let isPromise = p => (p instanceof Promise) || (typeof p === 'object' && p.then && p.catch);
```
Comment (voischev) [inline lib/posthtml.js]: It is named function at the bottom document

File: lib/posthtml.js [hunk 2]
```diff
@@ -23,31 +25,82 @@ class Posthtml {
...
+        let i = 0,
+            tryc = (a, b) => { let res; try { res = a(); } catch (e) { b(e); } return res; },
+            next = (res, cb) => {
+                // all plugins called
+                if (this.plugins.length <= i) {
+                    cb(null, res);
+                    return;
+                }
+
+                let _next = pluginResult => next(pluginResult || res, cb);
```
Comment (voischev) [inline lib/posthtml.js]: Incorrect indent

**Comment by voischev** [comment]
@zxqfox 
Write in Readme example use, describe options
 plz!

File: lib/posthtml.js [hunk 3]
```diff
@@ -23,31 +25,82 @@ class Posthtml {
...
             this.plugins.forEach(plugin => {
-                let result = plugin(tree);
-
-                if(result) {
-                    tree = result;
+                let result;
+                if (plugin.length === 2 || isPromise(result = plugin(tree))) {
+                    throw new Error(`Can't process synchronously because of async plugin: ${plugin.name}`);
                 }
+                tree = result || tree;
```
Comment (voischev) [inline lib/posthtml.js]: as it is not obviously get results. Maybe comments?)

**Comment by awinogradov** [comment]
:fire:

**Comment by zxqfox** [comment]
Need one more review here.

File: README.md [hunk 4]
```diff
@@ -77,9 +77,143 @@ gulp.task('html', function () {
...
+        return {
+            tag: 'span',
+            attrs: { class: 'img-wrapper' },
+            content: node
+        }
+    });
+
+    return tree;
+};
+```
```
Comment (voischev) [inline README.md]: javascript

File: README.md [hunk 5]
```diff
@@ -77,9 +77,143 @@ gulp.task('html', function () {
...
+### Classic asynchronous plugin example
+
+```
```
Comment (voischev) [inline README.md]: javascript

File: README.md [hunk 6]
```diff
@@ -77,9 +77,143 @@ gulp.task('html', function () {
...
+    function done() {
+        tasks -= 1;
+        if (!tasks) cb(null, tree);
+    }
+};
+```
+
+### Promised asynchronous plugin example
+
+```
```
Comment (voischev) [inline README.md]: javascript

**Comment by voischev** [comment]
Looks not bad. Do yourself a small edits

**Pull request merged**